stake-pool/single-pool/src/processor.rs

```diff
@@ -0,0 +1,1078 @@
+        ];
+        let stake_signers = &[&stake_seeds[..]];
+
+        let authority_seeds = &[
+            POOL_AUTHORITY_PREFIX,
+            validator_vote_info.key.as_ref(),
+            &[authority_bump_seed],
+        ];
+        let authority_signers = &[&authority_seeds[..]];
+
+        // create the pool stake account. user has alread transferred in rent plus the minimum
+        let minimum_delegation = stake::tools::get_minimum_delegation()?;
+        let stake_space = std::mem::size_of::<stake::state::StakeState>();
+        let stake_rent_plus_initial = rent
+            .minimum_balance(stake_space)
+            .saturating_add(minimum_delegation);
+
+        // XXX TODO FIXME jon raised the important issue that if we check != someone can dos pool creation
+        // unfortunately, that complicates the token minting, because we undermint if someone overpays
+        // i believe we should be able to load the stake account data after delegation to see the true stake amount
+        // TODO also write a test that overpayment goes smoothly
+        if pool_stake_info.lamports() < stake_rent_plus_initial {
+            return Err(SinglePoolError::WrongRentAmount.into());
+        }
+
+        let authorized = stake::state::Authorized::auto(pool_authority_info.key);
+
+        invoke_signed(
+            &system_instruction::allocate(pool_stake_info.key, stake_space as u64),
+            &[pool_stake_info.clone()],
+            stake_signers,
+        )?;
+
+        invoke_signed(
+            &system_instruction::assign(pool_stake_info.key, stake_program_info.key),
+            &[pool_stake_info.clone()],
+            stake_signers,
+        )?;
+
+        invoke_signed(
+            &stake::instruction::initialize_checked(pool_stake_info.key, &authorized),
+            &[
+                pool_stake_info.clone(),
+                rent_info.clone(),
+                pool_authority_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // delegate stake so it activates
+        invoke_signed(
+            &stake::instruction::delegate_stake(
+                pool_stake_info.key,
+                pool_authority_info.key,
+                validator_vote_info.key,
+            ),
+            &[
+                pool_stake_info.clone(),
+                validator_vote_info.clone(),
+                clock_info.clone(),
+                stake_history_info.clone(),
+                stake_config_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // mint tokens to the user corresponding to their stake deposit
+        // note that this is safe from economic attacks despite the stake being in an activating state
+        // because we block anyone else from depositing until it has activated
+        // XXX actually do we give a shit about the activating stake?
```

Comment:
I would say it's probably fine. Since there's no way to deactivate the stake, the worst is some confusion when someone creates a pool and can't deposit their activated stake immediately. You can add a helpful error message in deposit if someone tries to do that though